File: website/src/entities/connection.ts
export const DATABASE_ENGINES = ['postgresql', 'mysql', 'mariadb', 'sqlite', 'sqlserver', 'oracle', 'clickhouse', 'redis', 'mongodb'] as const
export type DatabaseEngine = (typeof DATABASE_ENGINES)[number]
export type AvailableDatabaseEngine = Extract<DatabaseEngine, 'postgresql' | 'mysql' | 'mariadb'>
export type ConnectionStatus = 'connected' | 'connecting' | 'disconnected' | 'error'
export type SSLMode = 'disable' | 'require' | 'verify-ca' | 'verify-full'

export type SSHTunnel = {
  enabled: boolean
  host: string
  port: number
  username: string
  privateKeyPath: string
  knownHostsPath: string
  hasPassword: boolean
}

export type Connection = {
  id: string
  workspaceId: string
  name: string
  engine: DatabaseEngine
  host: string
  port: number
  database: string
  username: string
  sslMode: SSLMode
  sslCaPath: string
  sslCertPath: string
  sslKeyPath: string
  proxyUrl: string
  sshTunnel: SSHTunnel
  readOnly: boolean
  autoReconnect: boolean
  maxOpenConns: number
  maxIdleConns: number
  connMaxLifetimeSeconds: number
  favorite: boolean
  hasPassword: boolean
  status: ConnectionStatus
  latencyMs?: number
  lastConnectedAt?: string
  createdAt: string
  updatedAt: string
}

export type ConnectionInput = {
  workspaceId: string
  name: string
  engine: AvailableDatabaseEngine
  host: string
  port: number
  database: string
  username: string
  password: string
  sslMode: SSLMode
  sslCaPath: string
  sslCertPath: string
  sslKeyPath: string
  proxyUrl: string
  sshTunnel: Omit<SSHTunnel, 'hasPassword'> & { password: string }
  readOnly: boolean
  autoReconnect: boolean
  maxOpenConns: number
  maxIdleConns: number
  connMaxLifetimeSeconds: number
}

export type CreateConnectionInput = ConnectionInput
export type UpdateConnectionInput = ConnectionInput

export type ConnectionTestResult = {
  ok: boolean
  latencyMs: number
  message: string
}

export type EngineCapability = {
  engine: DatabaseEngine
  label: string
  available: boolean
  capabilities: string[]
}

File: api/internal/core/entity/connection.go
package entity

import "time"

type Engine string

const (
	EnginePostgreSQL Engine = "postgresql"
	EngineMySQL      Engine = "mysql"
	EngineMariaDB    Engine = "mariadb"
	EngineSQLite     Engine = "sqlite"
	EngineSQLServer  Engine = "sqlserver"
	EngineOracle     Engine = "oracle"
	EngineClickHouse Engine = "clickhouse"
	EngineRedis      Engine = "redis"
	EngineMongoDB    Engine = "mongodb"
)

func (engine Engine) Valid() bool {
	return engine == EnginePostgreSQL || engine == EngineMySQL || engine == EngineMariaDB
}

type SSLMode string

const (
	SSLModeDisable    SSLMode = "disable"
	SSLModeRequire    SSLMode = "require"
	SSLModeVerifyCA   SSLMode = "verify-ca"
	SSLModeVerifyFull SSLMode = "verify-full"
)

type Connection struct {
	ID              string    `json:"id"`
	WorkspaceID     string    `json:"workspaceId"`
	Name            string    `json:"name"`
	Engine          Engine    `json:"engine"`
	Host            string    `json:"host"`
	Port            int       `json:"port"`
	Database        string    `json:"database"`
	Username        string    `json:"username"`
	PasswordCipher  string    `json:"-"`
	SSLMode         SSLMode   `json:"sslMode"`
	SSLCAPath       string    `json:"sslCaPath"`
	SSLCertPath     string    `json:"sslCertPath"`
	SSLKeyPath      string    `json:"sslKeyPath"`
	ProxyURL        string    `json:"proxyUrl"`
	SSHTunnel       SSHTunnel `json:"sshTunnel"`
	ReadOnly        bool      `json:"readOnly"`
	AutoReconnect   bool      `json:"autoReconnect"`
	MaxOpenConns    int       `json:"maxOpenConns"`
	MaxIdleConns    int       `json:"maxIdleConns"`
	ConnMaxLifetime int       `json:"connMaxLifetimeSeconds"`
	Favorite        bool      `json:"favorite"`
	CreatedAt       time.Time `json:"createdAt"`
	UpdatedAt       time.Time `json:"updatedAt"`
}

type SSHTunnel struct {
	Enabled        bool   `json:"enabled"`
	Host           string `json:"host"`
	Port           int    `json:"port"`
	Username       string `json:"username"`
	PasswordCipher string `json:"-"`
	Password       string `json:"-"`
	PrivateKeyPath string `json:"privateKeyPath"`
	KnownHostsPath string `json:"knownHostsPath"`
}

type EngineCapability struct {
	Engine       Engine   `json:"engine"`
	Label        string   `json:"label"`
	Available    bool     `json:"available"`
	Capabilities []string `json:"capabilities"`
}

func EngineCapabilities() []EngineCapability {
	return []EngineCapability{
		{Engine: EnginePostgreSQL, Label: "PostgreSQL", Available: true, Capabilities: []string{"connection", "explorer", "table-browser", "sql-editor", "schema", "operations"}},
		{Engine: EngineMySQL, Label: "MySQL", Available: true, Capabilities: []string{"connection", "explorer", "table-browser", "sql-editor", "schema", "operations"}},
		{Engine: EngineMariaDB, Label: "MariaDB", Available: true, Capabilities: []string{"connection", "explorer", "table-browser", "sql-editor", "schema", "operations"}},
		{Engine: EngineSQLite, Label: "SQLite", Available: false, Capabilities: []string{}},
		{Engine: EngineSQLServer, Label: "SQL Server", Available: false, Capabilities: []string{}},
		{Engine: EngineOracle, Label: "Oracle", Available: false, Capabilities: []string{}},
		{Engine: EngineClickHouse, Label: "ClickHouse", Available: false, Capabilities: []string{}},
		{Engine: EngineRedis, Label: "Redis", Available: false, Capabilities: []string{}},
		{Engine: EngineMongoDB, Label: "MongoDB", Available: false, Capabilities: []string{}},
	}
}

func (connection Connection) DefaultPort() int {
	if connection.Engine == EnginePostgreSQL {
		return 5432
	}
	return 3306
}
